src/client/index.js

```diff
@@ -0,0 +1,188 @@
+      //           utils.setUser(response.data);
+
+      //           return resolve(response);
+      //         });
+
+      //         socket[method]('authenticate', options);
+      //       };
+      //     });
+      //   }
+
+      //   if (app.io) {
+      //     connected(app.io, 'connected').then(handleAuth('emit')).then(function(response){
+
+      //     }).catch(function(error){
+      //       console.log('Errrr', error);
+      //     });
+      //   }
+
+      //   if (app.primus) {
+      //     connected(app.primus, 'open').then(handleAuth('send'));
+      //   }
+      // });
+
+      return new Promise(function(resolve, reject){
+        // TODO (EK): Handle OAuth logins
+        
+        // If we are using a REST client
+        if (app.rest) {
+          return app.service(endPoint).create(options).then(response => {
+            utils.setToken(response.token);
+            utils.setUser(response.data);
+
+            return resolve(response);
+          }).catch(reject);
+        }
+
+        if (app.io || app.primus) {
+          const transport = app.io ? 'io' : 'primus';
+          
+          app[transport].on('unauthorized', function(error) {
+            // console.error('Unauthorized', error);
+            return reject(error);
+          });
+
+          app[transport].on('authenticated', function (response) {
+            // console.log('authenticated', response);
+            utils.setToken(response.token);
+            utils.setUser(response.data);
+
+            return resolve(response);
+          });
+        }
+
+        // If we are using socket.io
+        if (app.io) {
+          // If we aren't already connected then throw an error
+          if (!app.io.connected) {
+            throw new Error('Socket not connected');
+          }
+          
+          app.io.on('disconnect', function(error) {
+            // console.error('Socket disconnected', error);
+            return reject(error);
+          });
+
+          app.io.emit('authenticate', options);
+        }
+
+        // If we are using primus
+        if (app.primus) {
+          // If we aren't already connected then throw an error
+          if (app.primus.readyState !== 3) {
+            throw new Error('Socket not connected');
+          }
+
+          app.primus.on('close', function(error) {
+            console.error('Socket disconnected', error);
+            return reject(error);
+          });
+          
+          app.primus.send('authenticate', options);
+        }
+      });
+    };
+
+    app.user = function() {
+      return utils.getUser();
+    };
+
+    app.logout = function() {
+      // remove user and token from localstorage
+      // on React native it's async storage
+      utils.clearToken();
+      utils.clearUser();
+    };
+
+    // Set up hook that adds adds token to data sent to server over sockets
+    app.mixins.push(function(service) {
+      service.before(hooks.populateParams());
```

Comment:
It looks like this needs hooks configured in the app for the authentication plugin to work at all. Should we throw an error if `service.before` and `service.after` isn't a function?